service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java

```diff
@@ -541,32 +424,71 @@ public PaymentSystemResponse saveFullOrderToDB(OrderResponseDto dto, String uuid
             dto.setPointsToUse(0);
         }
 
-        long sumToPayWithoutDiscountInCoins = formBagsToBeSavedAndCalculateOrderSum(bagsOrdered,
-            dto.getBags(), tariffsInfo);
+        long sumToPayWithoutDiscountInCoins =
+            formBagsToBeSavedAndCalculateOrderSum(bagsOrdered, dto.getBags(), tariffsInfo);
         checkIfUserHaveEnoughPoints(currentUser.getCurrentPoints(), dto.getPointsToUse());
         long sumToPayInCoins = reduceOrderSumDueToUsedPoints(sumToPayWithoutDiscountInCoins, dto.getPointsToUse());
+
         Order order = isExistOrder(dto, orderId);
         if (orderId != null) {
             checkIsOrderOfCurrentUser(currentUser, order);
+            if (order.getOrderStatus() != OrderStatus.FORMED
+                || order.getOrderPaymentStatus() != OrderPaymentStatus.UNPAID) {
+                throw new IllegalStateException(
+                    "Cannot update order: status must be FORMED and payment must be UNPAID");
+            }
+
+            entityManager.clear();
+
+            currentUser = userRepository.findByUuid(uuid);
+            order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new NotFoundException("Order not found after clearing context"));
```

Comment:
I do not think we should include `after clearing context` in the message, as it will be sent to the FE and it might not be clear what `context` refers to.

Reply:
please add logging here if you need